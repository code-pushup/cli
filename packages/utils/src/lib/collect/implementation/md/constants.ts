export const NEW_LINE = '\n';
