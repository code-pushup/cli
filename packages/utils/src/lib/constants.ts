export const SCORE_COLOR_RANGE = {
  GREEN_MIN: 0.9,
  YELLOW_MIN: 0.5,
};
