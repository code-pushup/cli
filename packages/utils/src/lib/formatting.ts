import {
  MAX_DESCRIPTION_LENGTH,
  MAX_ISSUE_MESSAGE_LENGTH,
  MAX_TITLE_LENGTH,
} from '@code-pushup/models';

export function slugify(text: string): string {
  return text
    .trim()
    .toLowerCase()
    .replace(/\s+|\//g, '-')
    .replace(/[^a-z\d-]/g, '');
}

export function pluralize(text: string): string {
  if (text.endsWith('y')) {
    return `${text.slice(0, -1)}ies`;
  }
  if (text.endsWith('s')) {
    return `${text}es`;
  }
  return `${text}s`;
}

export function formatBytes(bytes: number, decimals = 2) {
  const positiveBytes = Math.max(bytes, 0);

  // early exit
  if (positiveBytes === 0) {
    return '0 B';
  }

  const k = 1024;
  const dm = decimals < 0 ? 0 : decimals;
  const sizes = ['B', 'kB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB'];

  const i = Math.floor(Math.log(positiveBytes) / Math.log(k));

  return `${Number.parseFloat((positiveBytes / Math.pow(k, i)).toFixed(dm))} ${
    sizes[i]
  }`;
}

export function pluralizeToken(token: string, times: number = 0): string {
  return `${times} ${Math.abs(times) === 1 ? token : pluralize(token)}`;
}

export function formatDuration(duration: number): string {
  if (duration < 1000) {
    return `${duration} ms`;
  }
  return `${(duration / 1000).toFixed(2)} s`;
}

export function truncateText(text: string, maxChars: number): string {
  if (text.length <= maxChars) {
    return text;
  }
  const ellipsis = '...';
  return text.slice(0, maxChars - ellipsis.length) + ellipsis;
}

export function truncateTitle(text: string): string {
  return truncateText(text, MAX_TITLE_LENGTH);
}

export function truncateDescription(text: string): string {
  return truncateText(text, MAX_DESCRIPTION_LENGTH);
}

export function truncateIssueMessage(text: string): string {
  return truncateText(text, MAX_ISSUE_MESSAGE_LENGTH);
}
