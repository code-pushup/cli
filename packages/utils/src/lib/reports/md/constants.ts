export const NEW_LINE = '\n';
export const TAB = ' ';
