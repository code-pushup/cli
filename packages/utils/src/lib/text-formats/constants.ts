export const NEW_LINE = '\n';
export const TAB = '  ';
export const SPACE = ' ';
