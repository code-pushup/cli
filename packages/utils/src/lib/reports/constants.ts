import { TableHeading } from '@code-pushup/models';
import { Alignment } from './md';

// https://stackoverflow.com/questions/4651012/why-is-the-default-terminal-width-80-characters/4651037#4651037
export const TERMINAL_WIDTH = 80;

/* eslint-disable no-magic-numbers */
export const SCORE_COLOR_RANGE = {
  GREEN_MIN: 0.9,
  YELLOW_MIN: 0.5,
};
/* eslint-enable no-magic-numbers */

export const CATEGORIES_TITLE = 'üè∑ Categories';
export const FOOTER_PREFIX = 'Made with ‚ù§ by'; // replace ‚ù§Ô∏è with ‚ù§, because ‚ù§Ô∏è has output issues
export const CODE_PUSHUP_DOMAIN = 'code-pushup.dev';
export const README_LINK = 'https://github.com/code-pushup/cli#readme';
export const reportHeadlineText = 'Code PushUp Report';
export const reportOverviewTableHeaders = [
  {
    key: 'category',
    label: 'üè∑ Category',
  },
  {
    key: 'score',
    label: '‚≠ê Score',
  },
  {
    key: 'audits',
    label: 'üõ° Audits',
  },
] as const satisfies TableHeading[];
export const reportOverviewTableAlignment: Alignment[] = ['l', 'c', 'c'];
export const reportRawOverviewTableHeaders = ['Category', 'Score', 'Audits'];
export const reportMetaTableHeaders = [
  {
    key: 'commit',
    label: 'Commit',
  },
  {
    key: 'version',
    label: 'Version',
  },
  {
    key: 'duration',
    label: 'Duration',
  },
  {
    key: 'plugins',
    label: 'Plugins',
  },
  {
    key: 'categories',
    label: 'Categories',
  },
  {
    key: 'audits',
    label: 'Audits',
  },
] as const satisfies TableHeading[];
export const reportMetaTableAlignment: Alignment[] = [
  'l',
  'c',
  'c',
  'c',
  'c',
  'c',
];
export const pluginMetaTableHeaders = [
  {
    key: 'plugin',
    label: 'Plugin',
  },
  {
    key: 'audits',
    label: 'Audits',
  },
  {
    key: 'version',
    label: 'Version',
  },
  {
    key: 'duration',
    label: 'Duration',
  },
] as const satisfies TableHeading[];
export const pluginMetaTableAlignment: Alignment[] = ['l', 'c', 'c', 'c'];
// details headers

export const issuesTableHeadings = [
  {
    key: 'severity',
    label: 'Severity',
  },
  {
    key: 'message',
    label: 'Message',
  },
  {
    key: 'file',
    label: 'Source file',
  },
  {
    key: 'line',
    label: 'Line(s)',
  },
] as const satisfies TableHeading[];
