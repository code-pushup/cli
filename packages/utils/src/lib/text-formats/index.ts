import { details } from './html/details';
import {
  bold as boldHtml,
  code as codeHtml,
  italic as italicHtml,
} from './html/font-style';
import { link as linkHtml } from './html/link';
import { table as tableHtml } from './html/table';
import {
  bold as boldMd,
  code as codeMd,
  italic as italicMd,
  strikeThrough as strikeThroughMd,
} from './md/font-style';
import { h, h1, h2, h3, h4, h5, h6, headline } from './md/headline';
import { image } from './md/image';
import { link as linkMd } from './md/link';
import { indentation, li } from './md/list';
import { paragraphs } from './md/paragraphs';
import { lines, section } from './md/section';
import { table as tableMd } from './md/table';

export { NEW_LINE, SPACE, TAB } from './constants';
export { Hierarchy } from './types';

export const md = {
  bold: boldMd,
  italic: italicMd,
  strikeThrough: strikeThroughMd,
  code: codeMd,
  link: linkMd,
  image,
  headline,
  h,
  h1,
  h2,
  h3,
  h4,
  h5,
  h6,
  indentation,
  lines,
  li,
  section,
  paragraphs,
  table: tableMd,
};

export const html = {
  bold: boldHtml,
  italic: italicHtml,
  code: codeHtml,
  link: linkHtml,
  details,
  table: tableHtml,
};
