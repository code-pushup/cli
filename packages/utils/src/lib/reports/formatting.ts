import { AuditReport, Table } from '@code-pushup/models';
import { Hierarchy, NEW_LINE, SPACE, md } from '../text-formats';

const { headline, lines, link, section, table } = md;

export function tableSection(
  tableData: Table,
  options?: {
    level?: Hierarchy | 0;
  },
) {
  if (tableData.rows.length === 0) {
    return '';
  }
  const { level = 4 } = options ?? {};
  // if hierarchy is 0 do not apply heading styles
  const render = (h: string, l: Hierarchy | 0) =>
    l === 0 ? h : headline(h, l);
  return lines(
    tableData.title && render(tableData.title, level),
    table(tableData),
  );
}

// @TODO extract `Pick<AuditReport, 'docsUrl' | 'description'>` to a reusable schema and type
export function metaDescription({
  docsUrl,
  description,
}: Pick<AuditReport, 'docsUrl' | 'description'>): string {
  if (docsUrl) {
    const docsLink = link(docsUrl, 'ðŸ“– Docs');
    if (!description) {
      return section(docsLink);
    }
    const parsedDescription = description.toString().endsWith('```')
      ? `${description}${NEW_LINE + NEW_LINE}`
      : `${description}${SPACE}`;
    return section(`${parsedDescription}${docsLink}`);
  }
  if (description && description.trim().length > 0) {
    return section(description);
  }
  return '';
}
