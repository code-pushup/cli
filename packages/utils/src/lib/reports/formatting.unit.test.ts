import { describe, expect, it } from 'vitest';
import { NEW_LINE } from '../text-formats/constants';
import { metaDescription, tableSection } from './formatting';

describe('tableSection', () => {
  it('should render complete section', () => {
    expect(
      tableSection(
        {
          columns: [
            { key: 'phase', label: 'Phase' },
            { key: 'percentageLcp', label: '% of LCP', align: 'left' },
            { key: 'timing', label: 'Timing', align: 'right' },
          ],
          rows: [
            {
              phase: 'TTFB',
              percentageLcp: '27%',
              timing: '620 ms',
            },
            {
              phase: 'Load Delay',
              percentageLcp: '25%',
              timing: '580 ms',
            },
            {
              phase: 'Load Time',
              percentageLcp: '41%',
              timing: '940 ms',
            },
            {
              phase: 'Render Delay',
              percentageLcp: '6%',
              timing: '140 ms',
            },
          ],
        },
        { heading: 'LCP Breakdown', level: 3 },
      ),
    ).toMatchSnapshot();
  });
});

describe('metaDescription', () => {
  it('should return empty string if no options are given', () => {
    expect(metaDescription({})).toBe('');
  });

  it('should return description if only description is given', () => {
    expect(
      metaDescription({
        description: 'Audit to track bundle size',
      }),
    ).toBe(`Audit to track bundle size${NEW_LINE}`);
  });

  it('should return docsUrl if only docsUrl is given', () => {
    expect(
      metaDescription({
        docsUrl: `http://code-pushup.dev/audits/#lcp`,
      }),
    ).toBe(`[ðŸ“– Docs](http://code-pushup.dev/audits/#lcp)${NEW_LINE}`);
  });

  it('should docs and description if both given', () => {
    expect(
      metaDescription({
        description: 'Audit for loading performance',
        docsUrl: 'http://code-pushup.dev/audits/#lcp',
      }),
    ).toBe(
      `Audit for loading performance [ðŸ“– Docs](http://code-pushup.dev/audits/#lcp)${NEW_LINE}`,
    );
  });

  it('should include additional empty lines if description ends with a code block', () => {
    expect(
      metaDescription({
        description: 'Audit to loading performance```',
        docsUrl: 'http://code-pushup.dev/audits/#lcp',
      }),
    ).toBe(
      `Audit to loading performance\`\`\`${NEW_LINE}${NEW_LINE}[ðŸ“– Docs](http://code-pushup.dev/audits/#lcp)${NEW_LINE}`,
    );
  });
});
