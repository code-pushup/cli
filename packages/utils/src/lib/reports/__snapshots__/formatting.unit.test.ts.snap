// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tableSection > should render complete section 1`] = `
"### LCP Breakdown

|Phase|% of LCP|Timing|
|:--:|:--|--:|
|TTFB|27%|620 ms|
|Load Delay|25%|580 ms|
|Load Time|41%|940 ms|
|Render Delay|6%|140 ms|
"
`;
